man/kmrwrapper.1
'\"
.\" Copyright (C) 2012-2018 RIKEN R-CCS
.\" nroff -u0 -Tlp -man
.TH kmrwrapper.py 1 "2014-02-04" "KMR" "User Commands"
.SH NAME
kmrwrapper.py \- combines scripts of file split and job-script generation
.SH SYNOPSIS
.LP
.nf
\fBkmrwrapper.py\fR [-n num] [-e nodes] [-s pattern] [-p prefix] [-o outfile] [-d dir] [-O outdir] [-t rsctime] [-S sched] [-w file] [-f] -m mapper -r reducer file
.fi
.SH DESCRIPTION
.sp
.LP
kmrwrapper.py splits a file and generates a job-script.  It wraps
programs of "kmrfsplit.py" and "kmrgenscript.py".
.SH OPTIONS
.sp
.LP
The following options are supported:
.sp
.ne 2
.na
-n num
.ad
.sp .6
.RS 4n
Specifies the number of split.  Default is 1.
.RE
.sp
.ne 2
.na
-e nodes
.ad
.sp .6
.RS 4n
Specifies the number of processes to execute the job.  Default is 1.
.RE
.sp
.ne 2
.na
-s pattern
.ad
.sp .6
.RS 4n
Specifies a pattern string to split an input file.  It accepts a
regular expression.  Default is "\\n".
.RE
.sp
.ne 2
.na
-p prefix
.ad
.sp .6
.RS 4n
Specifies a prefix of input file names.  Default is "part".
.RE
.sp
.ne 2
.na
-o outfile
.ad
.sp .6
.RS 4n
Specifies a prefix of output file names.  Default is "output".
.RE
.sp
.ne 2
.na
-d indir
.ad
.sp .6
.RS 4n
Specifies an input directory.  Default is "./work".
.RE
.sp
.ne 2
.na
-O outdir
.ad
.sp .6
.RS 4n
Specifies an output directory.  It specifies the directory to hold the
job results.  Default is "./outdir".
.RE
.sp
.ne 2
.na
-t rsctime
.ad
.sp .6
.RS 4n
Specifies time limit in the job resources.  This is given by
"00:00:00" format.  Default is "00:10:00".
.RE
.sp
.ne 2
.na
-S sched
.ad
.sp .6
.RS 4n
Specifies a scheduler.  Only "K" is supported.
.RE
.sp
.ne 2
.na
-w scrfile
.ad
.sp .6
.RS 4n
Specifies a script output file.  Default is STDOUT.
.RE
.sp
.ne 2
.na
-f
.ad
.sp .6
.RS 4n
Forces to create an output directory if it does not exist.
.RE
.sp


CHANGELOG.md
# CHANGELOG

The list is sketchy.

## 1.9 (2018-08-27)

- 1.9 is a minor update
- Move GitHub repository from pf-aics-riken to riken-rccs
- Change the copyright holder name: RIKEN AICS to RIKEN R-CCS

## 1.8.1 (2016-04-25)

- Add ex/kmreprun.py, an example of simple shell command invoker
- Improve support of Mac OS X
- Several small bug fixes

## 1.8.0 (2015-12-25)

- Add a function tracing option for KMRViz, a KMR visualizer
- Add a simple scan function, kmr_scan_on_value()
- Add a load-leveling shuffle, kmr_shuffle_leveling_pair_count()
- Add a new kmr example, pagerank.c
- BUGFIX: kmr_match(), kmr_histogram_count_by_ranks(), kmr_find_key()

## 1.7.1 (2015-06-29)

- BUGFIX: kmr_take_one()
- BUGFIX: kmrrungenscript.py

## 1.7 (2015-06-22)

- Change license to BSD
- Strictly check options passed to kmr functions
- Add support for FOCUS supercomputer in kmrrungenscript.py
- Add Python API

## 1.6 (2015-04-01)

- Disable keep_open option in kmr functions except mapper functions
- Support no-fsync mode in Checkpoint/Restart
- Improve performance of KMRRUN when many input files are given
- BUGFIX: Checkpoint/Restart
- EXPERIMENTAL: Support selective mode that reduces IO overhead in
  Checkpoint/Restart

## 1.5 (2015-02-04)

- Reduce overhead of KMRRUN
- Temporal files for reducers are generated in subdirectories when
  using KMRRUN
- WORKAROUND: Abort on calling kmr_read_file_xxx with some MPI
  implementations
- WORKAROUND: Abort on calling kmr_map_processes and using KMRRUN
  with some Open MPI versions (1.6.3 - 1.8.1)

## 1.4 (2014-12-25)

- Resolved a scalability issue of KMRRUN and map functions that
  spawns child programs

## 1.3.2 (2014-11-07)

- Improve job script generator for KMRRUN

## 1.3.1 (2014-11-04)

- Add a word counting sample of KMRRUN
- BUGFIX: KMRRUN command line option parsing

## 1.3 (2014-09-17)

- Add a new function for affinity-aware file assignment

## 1.2 (2014-06-27)

- Support Checkpoint/Restart
- Add KMRRUN command that executes simple MapReduce workflow

## 1.1 (2013-09-20)

- 1.1 is a minor update

## 1.0 (2013-04-26)

- First release


cmd/kmrwrapper.py
#!/usr/bin/python
# Copyright (C) 2012-2018 RIKEN R-CCS

## \file kmrwrapper.py KMR-Shell File Spliter and Job-Script Generator.

import sys
import os
from optparse import OptionParser
from kmrfsplit import *
from kmrgenscript import *

## Wrapper script of kmrfsplit and kmrgenscript.
#  It works on Python 2.4 or later.

if __name__ == "__main__":

    usage = "usage: %prog [options] -m mapper -r reducer inputfile"
    parser = OptionParser(usage)

    parser.add_option("-n",
                      "--number-of-separation",
                      dest="nums",
                      type="int",
                      help="number of file separate",
                      metavar="number",
                      default=1)

    parser.add_option("-e",
                      "--number-of-exec-nodes",
                      dest="nodes",
                      type="int",
                      help="number of execute nodes",
                      metavar="number",
                      default=1)

    parser.add_option("-s",
                      "--separator",
                      dest="sep",
                      type="string",
                      help="separator string",
                      metavar="'string'",
                      default='\n')

    parser.add_option("-p",
                      "--separate-file-prefix",
                      dest="prefix",
                      type="string",
                      help="separate filename prefix",
                      metavar="'string'",
                      default='part')

    parser.add_option("-o",
                      "--outputfile",
                      dest="outfile",
                      type="string",
                      help="output filename prefix",
                      metavar="'string'",
                      default='output')

    parser.add_option("-d",
                      "--workdir",
                      dest="workdir",
                      type="string",
                      help="work directory",
                      metavar="'string'",
                      default='./work')

    parser.add_option("-O",
                      "--outputdir",
                      dest="outdir",
                      type="string",
                      help="output directory",
                      metavar="'string'",
                      default='./outdir')

    parser.add_option("-t",
                      "--resource-time",
                      dest="rsctime",
                      type="string",
                      help="resource time",
                      metavar="'string'",
                      default='00:10:00')

    parser.add_option("-m",
                      "--mapper",
                      dest="mapper",
                      type="string",
                      help="mapper path",
                      metavar="'string'")

    parser.add_option("-r",
                      "--reducer",
                      dest="reducer",
                      type="string",
                      help="reducer path",
                      metavar="'string'")

    parser.add_option("-S",
                      "--scheduler",
                      dest="sched",
                      type="string",
                      help="scheduler (default is 'K')",
                      metavar="'string'",
                      default='K')

    parser.add_option("-w",
                      "--write-scriptfile",
                      dest="scrfile",
                      type="string",
                      help="script filename",
                      metavar="'string'")

    parser.add_option("-f",
                      "--force",
                      dest="force",
                      action="store_true",
                      help="force option",
                      default=False)

    (options, args) = parser.parse_args()

    # check parameters.

    if len(args) <> 1 :
        parser.error("missing parameter")
        sys.exit()

    inputfile = args[0]

    if options.nodes > options.nums :
        print 'Error: number of execute nodes must be less than or equal to number of file separation.'
        sys.exit()

    # If number of nodes < number of input files,
    #  set M option (attach multi files to a mapper) to True.
    if options.nodes < options.nums :
        multi = True
    else :
        multi = False

    if not os.path.exists(inputfile) :
        print 'Error: inputfile %s is not exist.' % inputfile
        sys.exit()

    if os.path.exists(options.workdir) :
        if not os.path.isdir(options.workdir) :
            print 'Error: "%s" is not directory.' % options.workdir
            sys.exit()
    else:
        if options.force :
            try:
                os.mkdir(options.workdir)
            except IOError:
                print 'Error: could not create "%s".' % options.workdir
                sys.exit()
        else:
            print 'Error: directory "%s" is not exist. create it or use -f option.' % options.workdir
            sys.exit()

    splitfile(options.nums, options.sep, options.workdir, options.prefix, inputfile)

    selectscheduler(options.nodes, options.prefix, options.outfile,
                    options.workdir, options.outdir, options.rsctime,
                    options.mapper, options.reducer, multi,
                    options.sched, options.scrfile)

# Copyright (C) 2012-2018 RIKEN R-CCS
# This library is distributed WITHOUT ANY WARRANTY.  This library can be
# redistributed and/or modified under the terms of the BSD 2-Clause License.
